File: packages/runtime/src/layers.ts
/**
 * CPU reference layers. Mirrors tooling/python/gpu_utils_training/layers.py op for op.
 *
 * Every function works on ONE sequence of `n` tokens stored row-major in a Float32Array
 * (`x[t * c + i]`), accumulates in doubles and rounds to float32 on store, like torch.
 * Weight layouts: dense `[in, out]` (`w[i * out + o]`), depthwise conv `[taps, c]`,
 * dilated block `w1[tap][in][out]` = `w1[(tap * h + i) * h + o]`.
 */

export function sigmoid(x: number): number {
  return 1 / (1 + Math.exp(-x));
}

/** In-place ReLU. */
export function relu(x: Float32Array): Float32Array {
  for (let i = 0; i < x.length; i++) if (x[i]! < 0) x[i] = 0;
  return x;
}

/** Sum of embedding rows per token; ids equal to `paddingId` contribute nothing. */
export function sparseEmbed(
  rows: ArrayLike<ArrayLike<number>>,
  table: Float32Array,
  dim: number,
  paddingId: number,
): Float32Array {
  const n = rows.length;
  const out = new Float32Array(n * dim);
  for (let t = 0; t < n; t++) {
    const row = rows[t]!;
    for (let s = 0; s < row.length; s++) {
      const id = row[s]!;
      if (id === paddingId) continue;
      for (let c = 0; c < dim; c++) out[t * dim + c]! += table[id * dim + c]!;
    }
  }
  return out;
}

/** `y = x @ w + b` for `n` rows: x `[n, dIn]`, w `[dIn, dOut]`, b `[dOut]`. */
export function dense(
  x: Float32Array,
  n: number,
  dIn: number,
  w: Float32Array,
  b: Float32Array,
  dOut: number,
): Float32Array {
  const out = new Float32Array(n * dOut);
  const acc = new Float64Array(dOut);
  for (let t = 0; t < n; t++) {
    for (let o = 0; o < dOut; o++) acc[o] = b[o]!;
    for (let i = 0; i < dIn; i++) {
      const v = x[t * dIn + i]!;
      if (v === 0) continue;
      for (let o = 0; o < dOut; o++) acc[o]! += v * w[i * dOut + o]!;
    }
    for (let o = 0; o < dOut; o++) out[t * dOut + o] = acc[o]!;
  }
  return out;
}

/** Inclusive scan `h_t = a_t * h_{t-1} + b_t` per channel; `reverse` runs from the last token. */
export function affineScan(
  a: Float32Array,
  b: Float32Array,
  n: number,
  c: number,
  reverse = false,
): Float32Array {
  const out = new Float32Array(n * c);
  for (let ch = 0; ch < c; ch++) {
    let state = 0;
    for (let step = 0; step < n; step++) {
      const t = reverse ? n - 1 - step : step;
      state = a[t * c + ch]! * state + b[t * c + ch]!;
      out[t * c + ch] = state;
    }
  }
  return out;
}

export interface BiScanWeights {
  fWa: Float32Array;
  fBa: Float32Array;
  fWu: Float32Array;
  fBu: Float32Array;
  bWa: Float32Array;
  bBa: Float32Array;
  bWu: Float32Array;
  bBu: Float32Array;
}

/**
 * Gated bidirectional affine scan: x `[n, dIn]` → `[n, 2 * hidden]` (forward ‖ backward).
 * Per direction `a = sigmoid(x Wa + ba)`, `u = tanh(x Wu + bu)`, `h = a * h_prev + (1 - a) * u`.
 */
export function biScan(
  x: Float32Array,
  n: number,
  dIn: number,
  hidden: number,
  w: BiScanWeights,
): Float32Array {
  const out = new Float32Array(n * 2 * hidden);
  for (const reverse of [false, true]) {
    const wa = reverse ? w.bWa : w.fWa;
    const ba = reverse ? w.bBa : w.fBa;
    const wu = reverse ? w.bWu : w.fWu;
    const bu = reverse ? w.bBu : w.fBu;
    const a = dense(x, n, dIn, wa, ba, hidden);
    const u = dense(x, n, dIn, wu, bu, hidden);
    for (let i = 0; i < a.length; i++) {
      const g = sigmoid(a[i]!);
      a[i] = g;
      u[i] = (1 - g) * Math.tanh(u[i]!);
    }
    const h = affineScan(a, u, n, hidden, reverse);
    const off = reverse ? hidden : 0;
    for (let t = 0; t < n; t++)
      for (let ch = 0; ch < hidden; ch++) out[t * 2 * hidden + off + ch] = h[t * hidden + ch]!;
  }
  return out;
}

/** Per-channel convolution over time, zero padded, centre tap at `taps >> 1`; w `[taps, c]`. */
export function depthwiseConv(
  x: Float32Array,
  n: number,
  c: number,
  w: Float32Array,
  b: Float32Array,
  taps: number,
): Float32Array {
  const out = new Float32Array(n * c);
  const half = taps >> 1;
  for (let t = 0; t < n; t++) {
    for (let ch = 0; ch < c; ch++) {
      let acc = b[ch]!;
      for (let k = 0; k < taps; k++) {
        const src = t + k - half;
        if (src >= 0 && src < n) acc += w[k * c + ch]! * x[src * c + ch]!;
      }
      out[t * c + ch] = acc;
    }
  }
  return out;
}

/** `y = x + relu(conv3_dilated(x)) @ w2 + b2`; w1 `[3, h, h]`, w2 `[h, h]`. */
export function dilatedResidualBlock(
  x: Float32Array,
  n: number,
  h: number,
  w1: Float32Array,
  b1: Float32Array,
  w2: Float32Array,
  b2: Float32Array,
  dilation: number,
): Float32Array {
  const hid = new Float32Array(n * h);
  const acc = new Float64Array(h);
  for (let t = 0; t < n; t++) {
    for (let o = 0; o < h; o++) acc[o] = b1[o]!;
    for (let tap = 0; tap < 3; tap++) {
      const q = t + (tap - 1) * dilation;
      if (q < 0 || q >= n) continue;
      for (let i = 0; i < h; i++) {
        const v = x[q * h + i]!;
        if (v === 0) continue;
        const wo = (tap * h + i) * h;
        for (let o = 0; o < h; o++) acc[o]! += v * w1[wo + o]!;
      }
    }
    for (let o = 0; o < h; o++) hid[t * h + o] = Math.max(acc[o]!, 0);
  }
  const y = dense(hid, n, h, w2, b2, h);
  for (let i = 0; i < y.length; i++) y[i] = x[i]! + y[i]!;
  return y;
}

/** Mean over tokens: `[n, c]` → `[c]` (zeros when n = 0). */
export function meanPool(x: Float32Array, n: number, c: number): Float32Array {
  const out = new Float32Array(c);
  if (n === 0) return out;
  const acc = new Float64Array(c);
  for (let t = 0; t < n; t++) for (let ch = 0; ch < c; ch++) acc[ch]! += x[t * c + ch]!;
  for (let ch = 0; ch < c; ch++) out[ch] = acc[ch]! / n;
  return out;
}

/** `[n, a] ‖ [n, b]` → `[n, a + b]`; `right` may be a single `[b]` vector broadcast to every row. */
export function concatRows(
  left: Float32Array,
  a: number,
  right: Float32Array,
  b: number,
  n: number,
): Float32Array {
  const out = new Float32Array(n * (a + b));
  const broadcast = right.length === b;
  for (let t = 0; t < n; t++) {
    out.set(left.subarray(t * a, (t + 1) * a), t * (a + b));
    const r = broadcast ? right : right.subarray(t * b, (t + 1) * b);
    out.set(r, t * (a + b) + a);
  }
  return out;
}

File: tooling/python/gpu_utils_training/layers.py
"""Reference layers shared by every gpu-utils model family.

packages/runtime/src/layers.ts mirrors this file op for op; the shapes and conventions
below are the contract between the two.

Conventions
-----------
* Activations are ``[B, T, C]`` (batch, tokens, channels); ``mask`` is ``[B, T]`` bool
  with True for real tokens. Padded positions never influence real ones, so a batched
  forward equals the per-sequence forward the TypeScript runtime performs.
* Dense weights are stored ``[in, out]`` and applied as ``x @ W + b`` (row-major, so a
  GPU thread per output channel reads ``W[i * out + o]`` coalesced across threads).
* Gated scan: ``a = sigmoid(x Wa + ba)``, ``u = tanh(x Wu + bu)``,
  ``h_t = a_t * h_{t-1} + (1 - a_t) * u_t`` with ``h_0 = 0``. ``a`` is the *forget/keep*
  gate; ``(1 - a)`` scales the bounded candidate so the state stays in ``[-1, 1]`` and is
  robust to int6 quantization. This is the single convention used by every family.
* Depthwise convolution weights are ``[taps, C]``, zero padded, centre tap at
  ``taps // 2``; dilated block weights are ``[3, in, out]`` (tap, in, out).
"""

from __future__ import annotations

import math

import torch
from torch import Tensor, nn
from torch.nn import functional as F

from .qat import QParam


def sparse_embed(rows: Tensor, table: Tensor, padding_id: int) -> Tensor:
    """Sum of embedding rows per token. rows [B, T, S] long, table [R, E] -> [B, T, E].

    Slots equal to ``padding_id`` contribute nothing (``padding_id`` may be ``R``, i.e. one
    past the table, which is the default for the families).
    """
    keep = rows != padding_id
    safe = rows.masked_fill(~keep, 0)
    emb = F.embedding(safe, table) * keep.unsqueeze(-1).to(table.dtype)
    return emb.sum(dim=2)


def affine_scan(a: Tensor, b: Tensor, mask: Tensor | None = None, reverse: bool = False) -> Tensor:
    """Inclusive scan ``h_t = a_t * h_{t-1} + b_t`` along dim 1 (Hillis-Steele doubling).

    a, b: ``[B, T, C]``. Padded positions (``mask`` False) become the identity map
    ``(a=1, b=0)`` so trailing padding never leaks into a reversed scan, and their output
    is zeroed. ``reverse=True`` scans from the last token towards the first.
    """
    if mask is not None:
        m = mask.unsqueeze(-1).to(a.dtype)
        a = a * m + (1.0 - m)
        b = b * m
    if reverse:
        a = a.flip(1)
        b = b.flip(1)
    n = a.shape[1]
    stride = 1
    while stride < n:
        a_prev = F.pad(a[:, :-stride], (0, 0, stride, 0), value=1.0)
        b_prev = F.pad(b[:, :-stride], (0, 0, stride, 0), value=0.0)
        b = a * b_prev + b
        a = a * a_prev
        stride *= 2
    h = b.flip(1) if reverse else b
    if mask is not None:
        h = h * mask.unsqueeze(-1).to(h.dtype)
    return h


def _normal(shape: tuple[int, ...], std: float) -> Tensor:
    return torch.randn(shape) * std


class Dense(nn.Module):
    """``y = x @ W + b`` with W ``[in, out]``."""

    def __init__(self, d_in: int, d_out: int, std: float | None = None) -> None:
        super().__init__()
        self.d_in, self.d_out = d_in, d_out
        self.w = QParam(_normal((d_in, d_out), std if std is not None else 1.0 / math.sqrt(d_in)))
        self.b = QParam(torch.zeros(d_out))

    def forward(self, x: Tensor) -> Tensor:
        return x @ self.w() + self.b()


class BiScan(nn.Module):
    """Gated bidirectional affine scan: ``[B, T, d_in] -> [B, T, 2 * hidden]`` (forward ‖ backward).

    Per direction: ``a = sigmoid(x Wa + ba)``, ``u = tanh(x Wu + bu)``,
    ``h_t = a_t * h_{t-1} + (1 - a_t) * u_t``. Gate biases are initialised on a ramp so the
    channels start with a spread of memory lengths (mined from gpu-view).
    """

    def __init__(self, d_in: int, hidden: int) -> None:
        super().__init__()
        self.d_in, self.hidden = d_in, hidden
        std = 1.0 / math.sqrt(d_in)
        for direction in ("f", "b"):
            setattr(self, f"{direction}_wa", QParam(_normal((d_in, hidden), std)))
            setattr(self, f"{direction}_ba", QParam(torch.linspace(0.0, 3.0, hidden)))
            setattr(self, f"{direction}_wu", QParam(_normal((d_in, hidden), std)))
            setattr(self, f"{direction}_bu", QParam(torch.zeros(hidden)))

    def direction(self, x: Tensor, mask: Tensor | None, reverse: bool) -> Tensor:
        p = "b" if reverse else "f"
        a = torch.sigmoid(x @ getattr(self, f"{p}_wa")() + getattr(self, f"{p}_ba")())
        u = torch.tanh(x @ getattr(self, f"{p}_wu")() + getattr(self, f"{p}_bu")())
        return affine_scan(a, (1.0 - a) * u, mask, reverse)

    def forward(self, x: Tensor, mask: Tensor | None = None) -> Tensor:
        return torch.cat([self.direction(x, mask, False), self.direction(x, mask, True)], dim=-1)

    def tensors(self, prefix: str) -> dict[str, Tensor]:
        out: dict[str, Tensor] = {}
        for d in ("f", "b"):
            for n in ("wa", "ba", "wu", "bu"):
                out[f"{prefix}.{d}.{n}"] = getattr(self, f"{d}_{n}").w
        return out


class DepthwiseConv(nn.Module):
    """Per-channel 1-D convolution over time with ``taps`` (default 5) zero-padded taps.

    ``w[k, c]`` multiplies ``x[t + k - taps // 2, c]``. Padded/out-of-range tokens read as 0.
    """

    def __init__(self, channels: int, taps: int = 5, std: float = 0.1) -> None:
        super().__init__()
        self.channels, self.taps = channels, taps
        self.w = QParam(_normal((taps, channels), std))
        self.b = QParam(torch.zeros(channels))

    def forward(self, x: Tensor, mask: Tensor | None = None) -> Tensor:
        if mask is not None:
            x = x * mask.unsqueeze(-1).to(x.dtype)
        half = self.taps // 2
        w = self.w()
        xp = F.pad(x, (0, 0, half, self.taps - 1 - half))
        y = self.b().expand_as(x)
        for k in range(self.taps):
            y = y + xp[:, k : k + x.shape[1]] * w[k]
        if mask is not None:
            y = y * mask.unsqueeze(-1).to(y.dtype)
        return y


class DilatedResidualBlock(nn.Module):
    """``y = x + relu(conv3_dilated(x)) @ W2 + b2`` with zero padding; output masked.

    ``w1`` is ``[3, hidden, hidden]`` (tap, in, out): tap 0 reads ``t - d``, tap 2 ``t + d``.
    """

    def __init__(self, hidden: int, dilation: int) -> None:
        super().__init__()
        self.hidden, self.dilation = hidden, dilation
        self.w1 = QParam(_normal((3, hidden, hidden), math.sqrt(1.0 / (3 * hidden))))
        self.b1 = QParam(torch.zeros(hidden))
        self.w2 = QParam(_normal((hidden, hidden), math.sqrt(0.5 / hidden)))
        self.b2 = QParam(torch.zeros(hidden))

    def forward(self, x: Tensor, mask: Tensor | None = None) -> Tensor:
        m = None if mask is None else mask.unsqueeze(-1).to(x.dtype)
        if m is not None:
            x = x * m
        d = self.dilation
        w1 = self.w1()
        xp = F.pad(x, (0, 0, d, d))
        n = x.shape[1]
        h = self.b1() + xp[:, :n] @ w1[0] + x @ w1[1] + xp[:, 2 * d :] @ w1[2]
        y = x + torch.relu(h) @ self.w2() + self.b2()
        return y if m is None else y * m

    def tensors(self, prefix: str) -> dict[str, Tensor]:
        return {
            f"{prefix}.w1": self.w1.w,
            f"{prefix}.b1": self.b1.w,
            f"{prefix}.w2": self.w2.w,
            f"{prefix}.b2": self.b2.w,
        }


def masked_mean(x: Tensor, mask: Tensor | None) -> Tensor:
    """Mean over the token axis of the real tokens: ``[B, T, C] -> [B, C]``."""
    if mask is None:
        return x.mean(dim=1)
    m = mask.unsqueeze(-1).to(x.dtype)
    return (x * m).sum(dim=1) / m.sum(dim=1).clamp(min=1.0)
